cloudinit/distros/__init__.py

```diff
@@ -818,20 +823,108 @@ def create_user(self, name, **kwargs):
+
+        if "hashed_passwd" in kwargs:
+            has_password = True
+            password_key = "hashed_passwd"
+            if kwargs["hashed_passwd"]:
+                # Set password if hashed password is provided and non-empty
+                self.set_passwd(name, kwargs["hashed_passwd"], hashed=True)
+            else:
+                blank_password = True
+
+        if pre_existing_user:
+            if not has_password:
+                if "passwd" in kwargs:
+                    password_key = "passwd"
+                    # Only "plain_text_passwd" and "hashed_passwd"
+                    # are valid for an existing user.
+                    LOG.warning(
+                        "'passwd' in user-data is ignored for existing "
+                        "user %s",
+                        name,
+                    )
 
-        # Default locking down the account.  'lock_passwd' defaults to True.
-        # lock account unless lock_password is False.
+                # As no password set for the existing user in user-data then
+                # check if the existing user's hashed password value in
+                # /etc/shadow is blank (whether locked or not).
+                cmd = [
+                    "grep",
+                    "-q",
+                    "-e",
+                    "^%s::" % name,
+                    "-e",
+                    "^%s:!:" % name,
+                    "/etc/shadow",
+                ]
+                try:
+                    subp.subp(cmd)
+                except subp.ProcessExecutionError as e:
+                    if e.exit_code == 1:
+                        # Exit code 1 means 'grep' didn't find empty password
+                        has_password = True
+                    else:
+                        util.logexc(
+                            LOG,
+                            "Failed to Check if user %s already has "
+                            "a password",
+                            name,
+                        )
+                        raise e
+        else:
+            if "passwd" in kwargs:
+                has_password = True
+                password_key = "passwd"
+                if not kwargs["passwd"]:
+                    blank_password = True
+
+        # Default locking down the account. 'lock_passwd' defaults to True.
+        # Lock account unless lock_password is False in which case unlock
+        # account as long as a password (blank or otherwise) was specified.
         if kwargs.get("lock_passwd", True):
             self.lock_passwd(name)
+        elif has_password:
+            # 'lock_passwd: False' and password explicitly set
+            if blank_password and not pre_existing_user:
+                LOG.debug(
+                    "Allowing unlocking empty password for %s based on empty"
+                    " '%s' in user-data",
+                    name,
+                    password_key,
+                )
+
+            self.unlock_passwd(name)
```

Comment:
This unlock is still being  called even if `blank_password and pre_existing_user.` I think we still don't want to unlock the password in this case.

I'm working on a couple of unittests for these paths so we can be more certain we aren't adding exposure to unlocking insecure accounts by default.

Reply:
I'm renaming some of the variables to make it clear what their purpose is, so "blank_password" becomes "ud_blank_password_specified" to indicate this is where user-data contains a passwd/plain_text_passwd/hashed_passwd value of "".

Reply:
@dermotbradley thanks on this. If we get a chance to get this updated with any changes/refactors please ping. I'd like to get this into 24.3 which is scheduled for the week of Aug 19th.

Reply:
@blackboxsw Ok, I've pushed further changes.

Still needed are tests for snappy and for the various BSDs - neither of those are currently tested at all in test_create_users.py.

I'd also like to get this in 24.3 if possible.